DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+# First class functions
+
+| Section        | Value                                  |
+|----------------|----------------------------------------|
+| DIP:           | 27                                     |
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP proposes a radically simple approach by deciding that all functions are first class functions. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = function ReturnType(Parameters) {
+  // Function body.
+};
```

Comment:
How will current overloading be handled?

Currently this is valid,

    void f(string s) { ... }
    void f(int n)    { ... }

But after rewriting with the rule looks like it won't work as it will have name collisions.

    enum f = function void(string) {...};
    enum f = function void(int)    {...};

Direct call with explicit parameters may be able to distinguish both the same way it works now. But a second order function won't have any clue on which to pick.